### PR TITLE
ddl: fix rollback reorganize partition left intermediate state

### LINKED ISSUE #51090: ALTER TABLE t PARTITION BY KEY(a) leaves the table in an intermediate state if DDL is cancelled
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
drop table if exists t;
CREATE TABLE t (
  a int unsigned NOT NULL AUTO_INCREMENT,
  b varchar(255) DEFAULT NULL,
  PRIMARY KEY (a),
  KEY (b)
);

insert into t values (1,1),(2,2),(3,3),(4,4);
-- 1k
insert into t select null, t2.a from t, t t1, t t2, t t3, t t4;
-- 1M
insert into t select null, t2.a from t, t t2;

alter table t partition by key(b) partitions 100;
-- Wait until it shows as running, start a new session
-- check with admin show ddl jobs;
-- then
admin cancel ddl jobs <JOB_ID from above>;

-- Then the table is left in an intermediate state:
show create table t;

```

### 2. What did you expect to see? (Required)
Table rollbacked as previous state, non-partitioned

### 3. What did you see instead (Required)
Table is in an intermediate state, technically partitioned, but only with one partition and partitioned by NONE()
```
tidb> show create table t\G
*************************** 1. row ***************************
       Table: t
Create Table: CREATE TABLE `t` (
  `a` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `b` varchar(255) DEFAULT NULL,
  PRIMARY KEY (`a`) /*T![clustered_index] CLUSTERED */,
  KEY `b` (`b`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=1730010
PARTITION BY NONE ()
(PARTITION `pFullTable` COMMENT 'Intermediate partition during ALTER TABLE ... PARTITION BY ...')
1 row in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

```
tidb_version(): Release Version: v8.0.0-alpha-235-ga29bbb5ca1
Edition: Community
Git Commit Hash: a29bbb5ca180177b532d05283970ae17fc866ee9
Git Branch: master
UTC Build Time: 2024-02-09 12:49:36
GoVersion: go1.21.0
Race Enabled: false
Check Table Before Drop: false
Store: unistore
```

### PR DESCRIPTION
**Pull request by jiyfhust**
### What problem does this PR solve?

Issue Number: close #51090

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/parser/model/model.go [hunk 1]
```diff
@@ -1319,6 +1319,14 @@ func (pi *PartitionInfo) HasTruncatingPartitionID(pid int64) bool {
 	return false
 }
 
+// ClearReorgIntermediateInfo remove intermediate information used during reorganize partition.
+func (pi *PartitionInfo) ClearReorgIntermediateInfo() {
```
Comment (zimulala) [inline pkg/parser/model/model.go:1323]: Please add a simple UT to model_test.go.

File: pkg/ddl/partition.go [hunk 2]
```diff
@@ -2004,18 +2004,20 @@ func (w *worker) onDropTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (
 			job.State = model.JobStateCancelled
 			return ver, err
 		}
-		if partInfo.DDLType != model.PartitionTypeNone {
+		// ALTER TABLE ... PARTITION BY
+		if partInfo.Type != model.PartitionTypeNone {
```
Comment (zimulala) [inline pkg/ddl/partition.go:2008]: Why change this check?

File: pkg/ddl/partition.go [hunk 3]
```diff
@@ -2004,18 +2004,20 @@ func (w *worker) onDropTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (
...
 			// Also remove anything with the new table id
-			physicalTableIDs = append(physicalTableIDs, tblInfo.Partition.NewTableID)
+			physicalTableIDs = append(physicalTableIDs, partInfo.NewTableID)
 			// Reset if it was normal table before
-			if tblInfo.Partition.Type == model.PartitionTypeNone {
+			if tblInfo.Partition.Type == model.PartitionTypeNone ||
+				tblInfo.Partition.DDLType == model.PartitionTypeNone {
```
Comment (zimulala) [inline pkg/ddl/partition.go:2013]: Ditto. Do we have tests for this?

File: pkg/parser/model/model.go (hunk 1 above)
Comment (jiyfhust) [reply pkg/parser/model/model.go:1323]: Test is added. https://github.com/pingcap/tidb/pull/51631/commits/34b8c3a172060e5c2b0b23feb45944b9237f1156

**Review by zimulala — APPROVED** [review]

**Comment by zimulala** [comment]
PTAL @tangenta

**Comment by jiyfhust** [comment]
PTAL @tangenta

**Review by bb7133 — APPROVED** [review]
LGTM

**Review by D3Hunter — APPROVED** [review]

**Pull request merged**